ext/drbbdup/drbbdup.c

```diff
@@ -959,6 +1113,163 @@ drbbdup_include_encoding(drbbdup_manager_t *manager, uintptr_t new_encoding)
+
+    instrlist_t *ilist = decode_as_bb(drcontext, dr_fragment_app_pc(tag));
+    app_pc pc = instr_get_app_pc(drbbdup_first_app(ilist));
+    ASSERT(pc != NULL, "pc cannot be NULL");
+
+    bool do_flush = false;
+
+    /* Get the missing case. */
+    uintptr_t new_encoding = drbbdup_get_tls_raw_slot_val(DRBBDUP_ENCODING_SLOT);
+
+    dr_rwlock_write_lock(rw_lock);
+    drbbdup_manager_t *manager =
+        (drbbdup_manager_t *)hashtable_lookup(&manager_table, pc);
+    ASSERT(manager != NULL, "manager cannot be NULL");
+    ASSERT(manager->enable_dup, "duplication should be enabled");
+    ASSERT(new_encoding != manager->default_case.encoding,
+           "unhandled encoding cannot be the default case");
+
+    /* Could have been turned off potentially by an other thread. */
+    if (manager->enable_dynamic_handling) {
+        /* Case already registered potentially by an other thread. */
+        if (!drbbdup_encoding_already_included(manager, new_encoding)) {
+            /* By default, do case gen. */
+            bool do_gen = true;
+            if (opts.allow_gen != NULL) {
+                do_gen =
+                    opts.allow_gen(drcontext, tag, ilist, new_encoding,
+                                   &(manager->enable_dynamic_handling), opts.user_data);
+            }
+            if (do_gen)
+                drbbdup_include_encoding(manager, new_encoding);
+
+            /* Flush only if a new case needs to be generated or
+             * dynamic handling has been disabled.
+             */
+            do_flush = do_gen || !manager->enable_dynamic_handling;
+            if (do_flush) {
+                /* Mark that flushing is happening for drbbdup. */
+                manager->is_gen = true;
+                /* Increment counter so manager won't get freed due to flushing. */
+                manager->ref_counter++;
+            }
+
+            /* XXX i#4134: statistics -- Add increment to keep track generated cases. */
+        }
+    }
+    /* Regardless of whether or not flushing is going to happen, redirection will
+     * always be performed.
+     */
+    drbbdup_prepare_redirect(&mcontext, manager, pc);
+
+    dr_rwlock_write_unlock(rw_lock);
+
+    instrlist_clear_and_destroy(drcontext, ilist);
+
+    /* Refresh hit counter. */
+    if (opts.hit_threshold > 0) {
+        uint hash = drbbdup_get_hitcount_hash((intptr_t)pc);
+        DR_ASSERT(pt->hit_counts[hash] == 0);
+        pt->hit_counts[hash] = opts.hit_threshold; /* Reset threshold. */
+    }
+
+    /* Delete bb fragment. */
+    if (do_flush) {
+        LOG(drcontext, DR_LOG_ALL, 2,
+            "%s Found new case! Going to flush bb with"
+            "pc %p to generate a copy to handle the new case.\n",
+            __FUNCTION__, pc);
+
+        /* No locks held upon fragment deletion. */
+        /* XXX i#3778: To include once we support for deleting specific fragments. */
+        /* dr_delete_shared_fragment(tag); */
```

Comment:
Added commented code here which goes against our practices... but I think we can make an exception here for clarity.

Reply:
Done.